### PR TITLE
Port SRUpp code to 3.0.0-dev; and change some default settings of SRU

### PR DESCRIPTION
**Pull request by taoleicn**
- [x] change some default values; remove make_backward_compatible()
- [x] port SRU++ code
- [x] add unit tests
- [x] add experiment code - enwik8
- [x] add experiment code - wiki103
- [x] add experiment code - billion word
- [-] support incremental state for E2E ASR --> won't do. to be added in a separate PR
- [x] enwik8 reproducibility (42M model complete training)
- [x] wt103 reproducibility (115M model complete training)
- [x] billion word reproducibility (not complete training)

### CODE REVIEW

**Comment by bratao** [comment]
SRU with attention is amazing!! Congratulations @taoleicn and team! 👏 👏 👏 👏 
I´m super excited about it. You cite a paper. Could you share a pre-print if possible?

**Comment by taoleicn** [comment]
hi @bratao , thank you. We will release a tech report very soon. Will keep you posted!

File: experiments/srupp_experiments/prepare_corpus.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+
+if __name__ == '__main__':
+    import argparse
```
Comment (hpasapp) [inline experiments/srupp_experiments/prepare_corpus.py:3]: can we move the imports to the top of the file please?

File: sru/__init__.py [hunk 2]
```diff
@@ -6,3 +6,4 @@
 """
 from sru.version import __version__  # noqa: F401
 from sru.modules import SRU, SRUCell  # noqa: F401
+from sru.modules import SRUpp, SRUppCell  # noqa: F401
```
Comment (hpasapp) [inline sru/__init__.py:9]:
```
from sru.version import __version__ 
from sru.modules import SRU, SRUCell
from sru.modules import SRUpp, SRUppCell

__all__ = ['__version__', 'SRU', 'SRUCell', 'SRUpp', 'SRUppCell']
```
(this way you dont need the `# noqa` markings, I believe).

File: test/sru/test_torchscript.py [hunk 3]
```diff
@@ -32,3 +32,36 @@ def test_all(bidirectional, rescale, proj, layer_norm):
...
     assert (h - h_).abs().max() <= eps
     assert (c - c_).abs().max() <= eps
+
+
+@pytest.mark.parametrize("bidirectional", [False, True])
+@pytest.mark.parametrize("proj", [2, 4])
+@pytest.mark.parametrize("layer_norm", [False, True])
+@pytest.mark.parametrize("normalize_after", [False, True])
+@pytest.mark.parametrize("attn_every_n_layers", [1, 2])
+def test_srupp(bidirectional, proj, layer_norm, normalize_after, attn_every_n_layers):
```
Comment (hpasapp) [inline test/sru/test_torchscript.py:42]: the parametrize looks good :)

File: experiments/srupp_experiments/train_enwik8.py [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+    if args.local_rank == 0:
+        log_path = "{}_{}".format(args.log, random.randint(1, 100))
+        train_writer = SummaryWriter(log_dir=log_path+"/train")
+        dev_writer = SummaryWriter(log_dir=log_path+"/dev")
+
+    # set up distributed training
+    torch.cuda.set_device(args.local_rank)
+    device = torch.device("cuda", args.local_rank)
+    torch.distributed.init_process_group(backend="nccl")
+    set_seed(1234)
```
Comment (hpasapp) [inline experiments/srupp_experiments/train_enwik8.py:168]: maybe seed should be a command-line option, so we can try with different seeds?

File: sru/modules.py [hunk 5]
```diff
@@ -119,7 +119,9 @@ def __init__(self,
             self.activation = 'tanh'
         self.amp_recurrence_fp16 = amp_recurrence_fp16
         self.normalize_after = normalize_after
-        self.weight_c_init = weight_c_init
+        self.weight_c_init = float(1.0)
```
Comment (hpasapp) [inline sru/modules.py:122]: Note: `1.0` is already a float:

```
In [1]: type(1.0)
Out[1]: float

In [2]: type(float(1.0))
Out[2]: float
```

File: sru/modules.py [hunk 6]
```diff
@@ -161,13 +163,7 @@ def __init__(self,
...
 
     def reset_parameters(self):
-        """Properly initialize the weights of SRU, following the same
-        recipe as:
-        Xavier init:
-            http://proceedings.mlr.press/v9/glorot10a/glorot10a.pdf
-        Kaiming init:
-            https://arxiv.org/abs/1502.01852
-
+        """Properly initialize the weights of SRU.
```
Comment (hpasapp) [inline sru/modules.py:166]: suggest removing `Properly`, since add no new information on what initialization scheme is being used.

File: sru/modules.py [hunk 7]
```diff
@@ -633,33 +611,545 @@ def reset_parameters(self):
...
+                                     scale_val,
+                                     mask_c,
+                                     mask_pad)
+
+        if self.layer_norm is not None and self.normalize_after:
+            h = self.layer_norm(h)
+        return h, c
+
+
+class SRUpp(nn.Module):
```
Comment (hpasapp) [inline sru/modules.py:926]: how does `SRUpp` relate to `SRU` module? My previous understanding is that SRUpp is SRU where the custom_module includes attention? Are there some additional details that mean that we cannot use the existing `SRU` module?

File: sru/modules.py (hunk 5 above)
Comment (taoleicn) [reply sru/modules.py:122]: i used to get torchscript compilation errors for SRU w/o explicit float() annotation.

File: sru/modules.py (hunk 7 above)
Comment (taoleicn) [reply sru/modules.py:926]: the forward interface changed since we need to support attention related arguments, such as attention mask. the constructor will take additional options too.

**Comment by taoleicn** [comment]
hi Jing (@jinpoon),

Right now the SRU++ [forward() interface](https://github.com/asappresearch/sru/blob/3.0.0-dev-tao/sru/modules.py#L1064) would return a tuple of three items:

1. the output hidden states `h`
2. the internal hidden states `c`
3. a dictionary as a memory bank to hold other states `memory_dict : Dict[str, List[Tensor]]`. At the moment it only holds a list of tensors that are inputs to each layer of SRU++ `x^(i)`; to support incremental states, we can put additional keys such as `query_states`, `value_states` whose values will be the list of query tensors or key tensors of each SRU++ layer. 

let me know what you think. (not sure if returning a dictionary data type would be a concern for speed / c++ serving)

File: experiments/srupp_experiments/prepare_corpus.py (hunk 1 above)
Comment (taoleicn) [reply experiments/srupp_experiments/prepare_corpus.py:3]: will do.

File: sru/__init__.py (hunk 2 above)
Comment (taoleicn) [reply sru/__init__.py:9]: will do.

File: experiments/srupp_experiments/train_enwik8.py (hunk 4 above)
Comment (taoleicn) [reply experiments/srupp_experiments/train_enwik8.py:168]: will do.

File: sru/modules.py (hunk 6 above)
Comment (taoleicn) [reply sru/modules.py:166]: will do.

**Review by hpasapp — APPROVED** [review]

**Comment by taoleicn** [comment]
thank you @hpasapp :) 🎉

**Comment by taoleicn** [comment]
@bratao the tech report is out: https://arxiv.org/abs/2102.12459
